### PR TITLE
Prototype for Correlation IDs (GSI-491)

### PR DESCRIPTION
**Pull request by TheByronHimes**
# Correlation IDs in Requests

## Overview
A Correlation ID is a unique ID used to track related requests and is primarily used for logging purposes. The ID is passed along with the requests via header information until the terminal request.

This PR introduces a prototype that demonstrates how correlation IDs can be implemented in our microservices with Kafka and HTTP requests. The general idea is to use a ContextVar to make a received correlation ID available across a service.
If an ID is found to be invalid or missing, a new one should be generated and that occurrence should be logged (probably as a warning).

### For HTTP requests:
When an HTTP request is received, a middleware function will examine the header for a valid correlation ID. If needed, a new one will be generated and assigned to the header.
Then, the ID will be assigned to the ContextVar before the request is finally passed on to the endpoint function.
Within the endpoint function, or elsewhere in the application during request execution, the ID can be referenced by looking at the ContextVar.

### For Kafka events:
When a Kafka event is received, the correlation ID is extracted from the header. The header is not accessible using the current Kafka providers, so hexkit will need to be updated to allow that. For the prototype, the correlation ID is passed through an extra field in the event payload, but this is only because of the aforementioned limitation. The correlation ID is read from the event header and then the ContextVar is set before executing any resulting logic.

### CODE REVIEW

File: src/pci/adapters/inbound/event_sub.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+                    correlation_id,
+                )
+
+            async with set_correlation_id(correlation_id):
+                log.info("Set context correlation ID to %s", correlation_id)
+
+                with open(validated_payload.file_id, "w", encoding="utf-8") as file:
+                    file.write(
+                        f"The name of this file is {validated_payload.file_id}\n{correlation_id}"
+                    )
```
Comment (KerstenBreuer) [inline src/pci/adapters/inbound/event_sub.py]: This is just a placeholder for some real action?

File: src/pci/adapters/inbound/fastapi_/utils.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        return True
+    except ValueError:
+        return False
+
+
+async def correlation_id_middleware(request: Request, call_next):
+    """Check for correlation ID in the request headers.
+
+    If the correlation ID doesn't exist or isn't valid, generate a new one.
+    Set the header value if needed.
```
Comment (KerstenBreuer) [inline src/pci/adapters/inbound/fastapi_/utils.py]: I think, it should raise an error if the correlation ID is invalid. That is a bug that should never happen unless something strange is going on that we need to immediately explore.

File: src/pci/context_vars.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+
+@asynccontextmanager
+async def set_correlation_id(correlation_id: str):
+    """Set the correlation ID."""
+    async with set_context_var(correlation_id_var, correlation_id):
+        yield
+
+
+def get_correlation_id() -> str:
```
Comment (KerstenBreuer) [inline src/pci/context_vars.py:59]: Maybe raise an exception if no correlation ID was set.

**Review by KerstenBreuer — COMMENTED** [review]
Thanks a lot for that nice proposal. I like the idea of context vars.

However, could you maybe check if they work even in the following scenario:
1. An async task A sets the correlation ID, does some things, and gives control back to the event loop
2. Another async task B start an sets a different correlation ID, does some things, and gives control back to the event loop
3. Task A is ready to proceed, but is the correlation ID still the same as in step 1?

Moreover, maybe we should take some time to discuss this suggestion:
> If an ID is found to be invalid or missing, a new one should be generated and that occurrence should be logged (probably as a warning).
Let's first come to a conclusion on where the IDs are generated and then decide what happens if there is no ID.

File: src/pci/adapters/inbound/event_sub.py (hunk 1 above)
Comment (TheByronHimes) [reply src/pci/adapters/inbound/event_sub.py]: Correct

**Comment by TheByronHimes** [comment]
@KerstenBreuer I've added the requested test case

File: src/pci/adapters/inbound/fastapi_/utils.py (hunk 2 above)
Comment (TheByronHimes) [reply src/pci/adapters/inbound/fastapi_/utils.py]: Changed as part of [381076e](https://github.com/ghga-de/prototype-correlation-id/pull/1/commits/381076e6ccb149619e407166e0d0a58d9b7809cb)

File: src/pci/context_vars.py (hunk 3 above)
Comment (TheByronHimes) [reply src/pci/context_vars.py:59]: It seems reasonable to assume that if you're using the function, then there should be a value and if no value is present, something has gone wrong. Updated in [eacba71](https://github.com/ghga-de/prototype-correlation-id/pull/1/commits/eacba71e1c0874313864fb3cef72cfc6f61d41e9)

**Review by KerstenBreuer — CHANGES_REQUESTED** [review]
Great, that convinces me that context vars are in deed a viable solution for the given problem. @Cito what do you think?

File: tests/test_calls.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+async def test_rest_call_without_correlation_id(
+    joint_fixture: JointFixture, monkeypatch: pytest.MonkeyPatch  # noqa: F811
+):
+    """Make sure a new ID is generated if a REST api request is received with a missing
+    correlation ID.
+
+    Have to create a new rest client since monkeypatching won't affect the dispatch
+    function of already-instantiated middleware.
+    """
+    monkeypatch.setattr(
```
Comment (Cito) [inline tests/test_calls.py]: Instead of monkey patching, couldn't you just send test requests with different values in the correlation id header?

**Review by Cito — APPROVED** [review]
Thanks for working on this proposal, I like it and agree that ContextVars are an adequate choice for this purpose. I have used thread locals for such purposes before, which is a similar concept for threaded apps.

File: tests/test_calls.py (hunk 4 above)
Comment (TheByronHimes) [reply tests/test_calls.py]: You're absolutely right 😵‍💫

**Review by KerstenBreuer — APPROVED** [review]

**Pull request merged**